File: src/main/java/de/twometer/amongus/util/Scheduler.java
package de.twometer.amongus.util;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;

public class Scheduler {

    private static class Task {

        private static final AtomicInteger idCounter = new AtomicInteger(0);

        private final int id;

        private final long time;

        private final Runnable runnable;

        private Task(long time, Runnable runnable) {
            this.id = idCounter.incrementAndGet();
            this.time = time;
            this.runnable = runnable;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            return id == ((Task) o).id;
        }

        @Override
        public int hashCode() {
            return id;
        }

    }

    private final List<Task> tasks = new CopyOnWriteArrayList<>();

    public Task run(Runnable runnable) {
        return runLater(0, runnable);
    }

    public Task runLater(long delay, Runnable runnable) {
        var task = new Task(System.currentTimeMillis() + delay, runnable);
        tasks.add(task);
        return task;
    }

    public void cancel(Task task) {
        tasks.remove(task);
    }

    public void update() {
        for (var task : tasks) {
            if (task.time < System.currentTimeMillis()) {
                task.runnable.run();
                tasks.remove(task);
                return;
            }
        }
    }

}

File: src/main/kotlin/de/twometer/amogus/concurrency/Scheduler.kt
package de.twometer.amogus.concurrency

import java.util.concurrent.atomic.AtomicInteger

class ScheduledTask(val time: Long, val runnable: Runnable) {

    companion object {
        private val counter = AtomicInteger(0)
    }

    val id: Int = counter.incrementAndGet()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ScheduledTask

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id
    }

}

open class Scheduler {
    private val tasks = ArrayList<ScheduledTask>()

    fun runNow(runnable: Runnable): ScheduledTask = runLater(0, runnable)

    fun runLater(delay: Long, runnable: Runnable): ScheduledTask {
        val task = ScheduledTask(System.currentTimeMillis() + delay, runnable)
        synchronized(tasks) {
            tasks.add(task)
        }
        return task
    }

    fun cancel(task: ScheduledTask) {
        synchronized(tasks) {
            tasks.remove(task)
        }
    }

    fun update() {
        val executableTasks = ArrayList<ScheduledTask>()
        synchronized(tasks) {
            val iterator = tasks.iterator()
            while (iterator.hasNext()) {
                val task = iterator.next()
                if (task.time < System.currentTimeMillis()) {
                    executableTasks.add(task)
                    iterator.remove()
                }
            }
        }
        executableTasks.forEach { it.runnable.run() }
    }

}